[sharded_tensor] enable gloo/mpi backend to be used in sharded tensor tests

wanchaol opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #66351
* #64481
* __->__ #65855

This adjusted our test base to support non-nccl backend like gloo/mpi, so that we could test sharding on CPU with gloo/mpi backend.

Differential Revision: [D31287162](https://our.internmc.facebook.com/intern/diff/D31287162/)

cc @pietern @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @SciPioneer @H-Huang

Commit:
[sharded_tensor] support gloo backend in tests

This adjusted our test base to support non-nccl backend like gloo, so that we could test sharding on CPU with gloo backend.

Differential Revision: [D31287162](https://our.internmc.facebook.com/intern/diff/D31287162/)

[ghstack-poisoned]

Commit:
Update on "[sharded_tensor] support gloo backend in tests"


This adjusted our test base to support non-nccl backend like gloo, so that we could test sharding on CPU with gloo backend.

Differential Revision: [D31287162](https://our.internmc.facebook.com/intern/diff/D31287162/)

cc pietern mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse @SciPioneer @H-Huang

[ghstack-poisoned]

bowangbj commented on torch/testing/_internal/distributed/_sharded_tensor/__init__.py:
```diff
@@ -15,13 +15,30 @@ class ShardedTensorTestBase(MultiProcessTestCase):
-            init_method=f"file://{self.file_name}",
-        )
+    def init_pg(self, backend="nccl"):
+        if backend == "nccl":
```
probably just 

if backend not in ["nccl", "gloo", "mpi"]  --> raise error 

and then 

dist.init_process_group(
                backend=backend,
                world_size=self.world_size,
                rank=self.rank,
                init_method=f"file://{self.file_name}",
            )

bowangbj reviewed (commented):
Thanks Wanchao for the fix, just some minior stying issue

pritamdamania87 reviewed (commented):
This PR just adds the ability to run the tests on different backends but doesn't actual enable running tests on different backends yet right? Are you planning to have a separate PR to do that or are you planning to update this PR to also run all tests with all backends?

Commit:
Update on "[sharded_tensor] support gloo/mpi backend in tests"


This adjusted our test base to support non-nccl backend like gloo/mpi, so that we could test sharding on CPU with gloo/mpi backend.

Differential Revision: [D31287162](https://our.internmc.facebook.com/intern/diff/D31287162/)

cc pietern mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse @SciPioneer @H-Huang

[ghstack-poisoned]

wanchaol commented:
> This PR just adds the ability to run the tests on different backends but doesn't actual enable running tests on different backends yet right? Are you planning to have a separate PR to do that or are you planning to update this PR to also run all tests with all backends?

good point, I think this just enable the ability as you said to unblock tests in init_from_local_shards improvement, I will create a follow up PR to make sure we run all tests with different backends, will also rename this PR

Commit:
Update on "[sharded_tensor] enable gloo/mpi backend to be used in sharded tensor tests"


This adjusted our test base to support non-nccl backend like gloo/mpi, so that we could test sharding on CPU with gloo/mpi backend.

Differential Revision: [D31287162](https://our.internmc.facebook.com/intern/diff/D31287162/)

cc pietern mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse @SciPioneer @H-Huang

[ghstack-poisoned]

Commit:
Update on "[sharded_tensor] enable gloo/mpi backend to be used in sharded tensor tests"


This adjusted our test base to support non-nccl backend like gloo/mpi, so that we could test sharding on CPU with gloo/mpi backend.

Differential Revision: [D31287162](https://our.internmc.facebook.com/intern/diff/D31287162/)

cc pietern mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse @SciPioneer @H-Huang

[ghstack-poisoned]

bowangbj reviewed (approved):
LGTM, thx Wanchao

Commit:
Update on "[sharded_tensor] enable gloo/mpi backend to be used in sharded tensor tests"


This adjusted our test base to support non-nccl backend like gloo/mpi, so that we could test sharding on CPU with gloo/mpi backend.

Differential Revision: [D31287162](https://our.internmc.facebook.com/intern/diff/D31287162/)

cc pietern mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse @SciPioneer @H-Huang

[ghstack-poisoned]
